---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 05-map-one-eqtl.md in _episodes_rmd/
title: "Mapping A Single Gene Expression Trait"
teaching: 30
exercises: 30
questions:
- "How do I map one gene expression trait?"
objectives:
- "QTL mapping of an expression data set"
keypoints:
- "To run a QTL analysis for expression data"
source: Rmd
---




### Load Libraries  


~~~
library(tidyverse)
library(qtl2)
library(qtl2convert)
library(GGally)
library(broom)
library(corrplot)
library(RColorBrewer)
library(knitr)
~~~
{: .language-r}


### Load Data

Let's load the data. This time we are loading gene expression for `21,771` genes.  


~~~
#expression data
load("../data/attie_DO500_expr.datasets.RData")
load("../data/dataset.islet.rnaseq.RData")

##phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")

##mapping data
load("../data/attie_DO500_mapping.data.RData")

probs = readRDS("../data/attie_DO500_genoprobs_v5.rds")
~~~
{: .language-r}


### Expression Data

Gene expression information is in the `norm` data object. These  counts have been normalised which has been explianed in a previous lesson [link???]. Because we are 
working with `insulin tAUC` phenotype, let's map the expression counts for 
`Hnf1b` which is known to influence this phenotype is these data. Before we do, 
let's check the distribution for `Hnf1b` expression data.

!! ADD here are about normalisation !!



~~~
hist(counts$ENSMUSG00000020679, main = "Hnf1b")
~~~
{: .language-r}

<img src="../fig/rmd-05-hist_untransformed-1.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" />

The histogram indicates that distribution of these counts are normalised.


### The Marker Map  

The marker map for each chromosome is stored in the `map` object. This is used 
to plot the LOD scores calculated at each marker during QTL mapping.  Here we 
are using the 69K grid marker file.  

We are using the same marker map as in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/05-review-mapping-steps/index.html#the-marker-map)


### Genotype probabilities  

We have already calculated genotype probabilities which we loaded above called 
`probs`.  This contains the 8 state genotype probabilities using the 69k grid 
map of the same 500 DO mice that also have clinical phenotypes. 

We have explored this earlier in th previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/05-review-mapping-steps/index.html#genotype-probabilities)


### [Kinship Matrix](https://smcclatchy.github.io/mapping/04-calc-kinship/)

The kinship matrix has already been calculated and loaded in above.  Again, we 
have explored this in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/05-review-mapping-steps/index.html#kinship-matrix)


### Covariates    

Now let's add the necessary covariates. For `Hnf1b` expression data, let's see 
which covariates are significant.


~~~
###merging covariate data and expression data to test for sex, wave and diet_days.

cov.counts <- merge(covar, norm, by=c("row.names"), sort=F)

#testing covairates on expression data

tmp = cov.counts %>%
        dplyr::select(mouse, sex, DOwave, diet_days, ENSMUSG00000020679) %>%
        gather(expression, value, -mouse, -sex, -DOwave, -diet_days) %>%
        group_by(expression) %>%
        nest()
mod_fxn = function(df) {
  lm(value ~ sex + DOwave + diet_days, data = df)
}
tmp = tmp %>%
  mutate(model = map(data, mod_fxn)) %>%
  mutate(summ = map(model, tidy)) %>%
  unnest(summ) 
#  kable(tmp, caption = "Effects of Sex, Wave & Diet Days on Expression")

tmp
~~~
{: .language-r}



~~~
# A tibble: 4 × 8
# Groups:   expression [1]
  expression         data     model  term      estimate std.e…¹ stati…²  p.value
  <chr>              <list>   <list> <chr>        <dbl>   <dbl>   <dbl>    <dbl>
1 ENSMUSG00000020679 <tibble> <lm>   (Interce… -1.70    0.506     -3.36 8.72e- 4
2 ENSMUSG00000020679 <tibble> <lm>   sexM      -0.199   0.0824    -2.42 1.60e- 2
3 ENSMUSG00000020679 <tibble> <lm>   DOwave     0.510   0.0370    13.8  3.73e-35
4 ENSMUSG00000020679 <tibble> <lm>   diet_days  0.00403 0.00386    1.04 2.97e- 1
# … with abbreviated variable names ¹​std.error, ²​statistic
~~~
{: .output}



~~~
tmp %>%
  filter(term != "(Intercept)") %>%
  mutate(neg.log.p = -log10(p.value)) %>%
  ggplot(aes(term, neg.log.p)) +
    geom_point() +
    facet_wrap(~expression) +
    labs(title = "Significance of Sex, Wave & Diet Days on Expression") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
rm(tmp)
~~~
{: .language-r}

<img src="../fig/rmd-05-covariates sig-1.png" alt="plot of chunk covariates sig" width="612" style="display: block; margin: auto;" />

We can see that sex, DOwave and diet are significant.  Here DOwave is the group 
or batch number as not all mice were submitted for genotyping at the same time.  
Because of this, we now have to correct for it.



~~~
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)
pheno_clin$diet_days = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave + diet_days, data = pheno_clin)
~~~
{: .language-r}

### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now let's perform the genome scan!


~~~
qtl = scan1(genoprobs = probs, 
            pheno = norm[,"ENSMUSG00000020679", drop = FALSE], 
            kinship = K, 
            addcovar = covar, 
            cores = 2)
~~~
{: .language-r}

Let's plot it


~~~
plot_scan1(x = qtl, 
           map = map, 
           lodcolumn = "ENSMUSG00000020679", 
           main = colnames(qtl))
abline(h = 6, col = 2, lwd = 2)
~~~
{: .language-r}

<img src="../fig/rmd-05-qtl_plot-1.png" alt="plot of chunk qtl_plot" width="576" style="display: block; margin: auto;" />


### [Finding LOD peaks](https://smcclatchy.github.io/mapping/07-find-lod-peaks/)

Let's find LOD peaks


~~~
lod_threshold = 6
peaks = find_peaks(scan1_output = qtl, 
                   map = map, 
                   threshold = lod_threshold, 
                   peakdrop = 4, prob = 0.95)
kable(peaks %>% 
        dplyr::select(-lodindex) %>% 
        arrange(chr, pos), caption = "Phenotype QTL Peaks with LOD >= 6")
~~~
{: .language-r}



Table: Phenotype QTL Peaks with LOD >= 6

|lodcolumn          |chr |       pos|       lod|     ci_lo|     ci_hi|
|:------------------|:---|---------:|---------:|---------:|---------:|
|ENSMUSG00000020679 |9   | 109.08150|  6.402594| 106.66772| 111.56048|
|ENSMUSG00000020679 |11  |  84.40138| 36.878942|  83.64714|  84.40138|
|ENSMUSG00000020679 |15  |  68.91018|  6.121254|  67.82232|  70.38004|
|ENSMUSG00000020679 |18  |  70.89669|  6.042135|  33.80428|  71.96869|


> ## Challenge
> Now choose another gene expression trait in `norm` data object and perform 
> the same steps.  
> 1). Check the distribution. Does it need transforming? 
> 2). Are there any sex, batch, diet effects? 
> 3). Run a genome scan with the genotype probabilities and kinship provided.  
> 4). Plot the genome scan for this gene.
> 5). Find the peaks above LOD score of 6.  
>
> > ## Solution
> > 
> > 
> {: .solution}
{: .challenge}
