---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 06-map-many-eqtls.md in _episodes_rmd/
title: "Mapping Many Gene Expression Traits"
teaching: 30
exercises: 30
questions:
- "How do I map many genes?"
objectives:
- "To map several genes at the same time"
source: Rmd
---



### Load Libraries  


~~~
library(tidyverse)
library(knitr)
library(broom)
library(qtl2)
library(qtl2ggplot)
library(RColorBrewer)

source("../code/gg_transcriptome_map.R")
source("../code/qtl_heatmap.R")
~~~
{: .language-r}

Before we begin this lesson, we need to create another directory called `results` in our main directory.  You can do this by clicking on the "Files" tab and navigate into the main directory.  Then select "New Folder" and name it "results".

### Load Data


~~~
# expression data
load("../data/attie_DO500_expr.datasets.RData")

# data from paper
load("../data/dataset.islet.rnaseq.RData")

# phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")

# mapping data
load("../data/attie_DO500_mapping.data.RData")

# genotype probabilities
probs = readRDS("../data/attie_DO500_genoprobs_v5.rds")
~~~
{: .language-r}

### Data Selection

For this lesson, lets choose a random set of 50 gene expression phenotypes.




~~~
genes = colnames(norm)

sams <- sample(length(genes), 50, replace = FALSE, prob = NULL)
genes <- genes[sams]
~~~
{: .language-r}

### Expression Data

Lets check the distributin of these 50 gene expression phenotypes


~~~
par(mfrow=c(3,4))
for(gene in genes[1:20]){
  hist(norm[,gene], main = dataset.islet.rnaseq$annots[dataset.islet.rnaseq$annots$gene_id == gene,]$symbol)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-hist_untransformed-1.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-hist_untransformed-2.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" />
The histogram indicates that distribution of these counts are normalised (as they should be).

### The Marker Map  

We are using the same marker map as in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#the-marker-map)


### Genotype probabilities  

We have explored this earlier in th previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#genotype-probabilities).  But, as a reminder, we have already calculated genotype probabilities which we loaded above called `probs`.  This contains the 8 state genotype probabilities using the 69k grid  map of the same 500 DO mice that also have clinical phenotypes. 


### [Kinship Matrix](https://smcclatchy.github.io/mapping/04-calc-kinship/)

We have explored the kinship matrix in the previous [lesson](https://smcclatchy.github.io/gene-expression-qtl/04-review-mapping-steps/index.html#kinship-matrix). It has already been calculated and loaded in above. 


### Covariates    

Now let's add the necessary covariates. For these 50 gene expression data, let's see which covariates are significant.


~~~
###merging covariate data and expression data to test for sex, wave and diet_days.

cov.counts <- merge(covar, norm[,genes], by=c("row.names"), sort=F)

#testing covairates on expression data

tmp = cov.counts %>%
        dplyr::select(mouse, sex, DOwave, diet_days, names(cov.counts[,genes])) %>%
        gather(expression, value, -mouse, -sex, -DOwave, -diet_days) %>%
        group_by(expression) %>%
        nest()
mod_fxn = function(df) {
  lm(value ~ sex + DOwave + diet_days, data = df)
}
tmp = tmp %>%
  mutate(model = map(data, mod_fxn)) %>%
  mutate(summ = map(model, tidy)) %>%
  unnest(summ) 
#  kable(tmp, caption = "Effects of Sex, Wave & Diet Days on Expression")

tmp
~~~
{: .language-r}



~~~
# A tibble: 200 × 8
# Groups:   expression [50]
   expression         data     model  term     estimate std.e…¹ stati…²  p.value
   <chr>              <list>   <list> <chr>       <dbl>   <dbl>   <dbl>    <dbl>
 1 ENSMUSG00000032068 <tibble> <lm>   (Interc… -5.04e-1 0.593    -0.850 3.96e- 1
 2 ENSMUSG00000032068 <tibble> <lm>   sexM     -8.35e-2 0.0966   -0.864 3.88e- 1
 3 ENSMUSG00000032068 <tibble> <lm>   DOwave    2.68e-1 0.0434    6.18  1.72e- 9
 4 ENSMUSG00000032068 <tibble> <lm>   diet_da… -9.28e-4 0.00453  -0.205 8.38e- 1
 5 ENSMUSG00000095375 <tibble> <lm>   (Interc…  3.55e-1 0.617     0.575 5.66e- 1
 6 ENSMUSG00000095375 <tibble> <lm>   sexM     -2.14e-1 0.101    -2.12  3.44e- 2
 7 ENSMUSG00000095375 <tibble> <lm>   DOwave   -6.60e-2 0.0452   -1.46  1.45e- 1
 8 ENSMUSG00000095375 <tibble> <lm>   diet_da… -5.38e-4 0.00472  -0.114 9.09e- 1
 9 ENSMUSG00000031889 <tibble> <lm>   (Interc…  1.34e+0 0.556     2.40  1.67e- 2
10 ENSMUSG00000031889 <tibble> <lm>   sexM     -6.42e-1 0.0907   -7.09  6.93e-12
# … with 190 more rows, and abbreviated variable names ¹​std.error, ²​statistic
~~~
{: .output}



~~~
tmp %>%
  filter(term != "(Intercept)") %>%
  mutate(neg.log.p = -log10(p.value)) %>%
  ggplot(aes(term, neg.log.p)) +
    geom_point() +
    facet_wrap(~expression, ncol=10) +
    labs(title = "Significance of Sex, Wave & Diet Days on Expression") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
rm(tmp)
~~~
{: .language-r}

<img src="../fig/rmd-06-covariates_sig-1.png" alt="plot of chunk covariates_sig" width="612" style="display: block; margin: auto;" />

We can see that `DOwave` is the most significant.  However, given that a few are influenced by `sex` and `diet_days`, we will have to correct for those as well. 



~~~
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)
pheno_clin$diet_days = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave + diet_days, data = pheno_clin)
~~~
{: .language-r}

### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now lets perform the genome scan!  We are also going to save our qtl results in an `Rdata` file to be used in further lessons. 

### QTL Scans


~~~
qtl.file = "../results/gene.norm_qtl_cis.trans.Rdata"

if(file.exists(qtl.file)) {
  load(qtl.file)
  } else {
    qtl = scan1(genoprobs = probs, 
                pheno = norm[,genes, drop = FALSE],
                kinship = K, 
                addcovar = covar, 
                cores = 2)
    save(qtl, file = qtl.file)
    }
~~~
{: .language-r}

### QTL plots




~~~
par(mfrow=c(3,4))
for(i in 1:20) {
  plot_scan1(x = qtl, 
             map = map, 
             lodcolumn = i, 
             main = colnames(qtl)[i])
  abline(h = 6, col = 2, lwd = 2)
  }
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_plots-1.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-06-qtl_plots-2.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" />

### QTL Peaks

We are also going to save our peak results so we can use these again else where.  First, lets get out peaks with a LOD score greater than 6. 


~~~
lod_threshold = 6
peaks = find_peaks(scan1_output = qtl, 
                   map = map, 
                   threshold = lod_threshold, 
                   peakdrop = 4, 
                   prob = 0.95)
~~~
{: .language-r}

We will save these peaks into a csv file. 


~~~
kable(peaks %>% 
        dplyr::select(-lodindex) %>% 
        arrange(chr, pos), caption = "Expression QTL (eQTL) Peaks with LOD >= 6")

write_csv(peaks, "../results/gene.norm_qtl_peaks_cis.trans.csv")
~~~
{: .language-r}


Table: Phenotype QTL Peaks with LOD >= 6

|lodcolumn          |chr |        pos|        lod|      ci_lo|     ci_hi|
|:------------------|:---|----------:|----------:|----------:|---------:|
|ENSMUSG00000033389 |1   |  24.116007|   6.296328|  18.001823|  32.48454|
|ENSMUSG00000041189 |1   |  38.681019|   8.009293|  37.690607|  38.99625|
|ENSMUSG00000015112 |1   |  74.157715|   6.390019|  73.792993| 190.01604|
|ENSMUSG00000058420 |1   | 190.002808|   6.531293| 135.481336| 191.89737|
|ENSMUSG00000026637 |1   | 192.137668|  13.183821| 191.655992| 192.51205|
|ENSMUSG00000019945 |2   |  72.077418|   6.609965|  17.713445| 169.28966|
|ENSMUSG00000081046 |2   | 129.184651|   6.497957|  12.915206| 135.45041|
|ENSMUSG00000097283 |2   | 153.419672|   8.793228| 150.991127| 153.83765|
|ENSMUSG00000027487 |2   | 156.136279|   6.608979| 154.085478| 156.75363|
|ENSMUSG00000027487 |2   | 161.576254|   7.091103| 161.309135| 162.89119|
|ENSMUSG00000054455 |2   | 165.445995|   8.888555| 163.582672| 166.47560|
|ENSMUSG00000097283 |2   | 178.833418|  30.515193| 176.416737| 178.90319|
|ENSMUSG00000027718 |3   |  33.606276|  31.217791|  32.304940|  33.65661|
|ENSMUSG00000027718 |3   |  37.165286|  75.126515|  36.852209|  37.32921|
|ENSMUSG00000027722 |3   |  37.374870|   6.214390|  34.262328| 114.46680|
|ENSMUSG00000048416 |3   |  67.791179|   7.134414|  66.785770|  73.56933|
|ENSMUSG00000000340 |3   | 114.948141|   9.484962| 111.607392| 114.97076|
|ENSMUSG00000037994 |3   | 135.232817|  49.079161| 135.232817| 135.23282|
|ENSMUSG00000021127 |4   |   3.798433|   6.132635|   3.000000| 149.94046|
|ENSMUSG00000070980 |4   |  54.766817|  11.218926|  54.527023|  55.60324|
|ENSMUSG00000096177 |4   |  65.427487|   8.792047|  63.528309|  78.76958|
|ENSMUSG00000033389 |4   | 111.238045|   6.214717| 109.522381| 116.12848|
|ENSMUSG00000050854 |4   | 118.471245|  35.618326| 118.397472| 118.54342|
|ENSMUSG00000029804 |4   | 143.211132|   6.035045| 138.750051| 148.34659|
|ENSMUSG00000029321 |4   | 154.874097|   6.069940|  20.129417| 156.21626|
|ENSMUSG00000081046 |5   |  98.997378|   6.773936|  92.854978| 103.17849|
|ENSMUSG00000029321 |5   | 103.614608|  21.562401| 103.586587| 105.67241|
|ENSMUSG00000099095 |5   | 108.046168|   6.777903| 102.956536| 112.75631|
|ENSMUSG00000032305 |6   |  37.349577|   6.720697|  16.376879|  48.92068|
|ENSMUSG00000071519 |6   |  41.406513|  17.315115|  40.920539|  41.40651|
|ENSMUSG00000002797 |6   |  54.997148|  39.222855|  54.855033|  55.18362|
|ENSMUSG00000029804 |6   |  58.488862|   8.546978|  40.993625|  60.76170|
|ENSMUSG00000089634 |6   |  86.133168|  11.672479|  85.420854|  86.55060|
|ENSMUSG00000048416 |6   | 125.226964|   6.154786|  45.277826| 125.94593|
|ENSMUSG00000031532 |6   | 125.553788|   6.312708|  63.210018| 126.65001|
|ENSMUSG00000070980 |6   | 127.078807|   6.130341| 126.722486| 135.05159|
|ENSMUSG00000080968 |6   | 133.867522|   6.004287| 129.390996| 139.21454|
|ENSMUSG00000015112 |6   | 134.185817|   6.123813| 129.110159| 135.65721|
|ENSMUSG00000070980 |6   | 145.839910|   6.202777| 145.391545| 146.31853|
|ENSMUSG00000025557 |7   |  36.613646|   6.506206|  35.767556| 133.15669|
|ENSMUSG00000038738 |7   |  44.445619|  19.010804|  43.844466|  44.86733|
|ENSMUSG00000080383 |7   |  57.912994|   6.407887|  50.696529| 101.62734|
|ENSMUSG00000058420 |7   | 118.237771|  29.916509| 117.782005| 118.45922|
|ENSMUSG00000021676 |8   |  22.086620|   6.149473|  18.457359| 115.32557|
|ENSMUSG00000031532 |8   |  33.827024|  17.787609|  33.095603|  34.20777|
|ENSMUSG00000002797 |8   |  95.702894|   6.560175|  94.937120| 103.87985|
|ENSMUSG00000049307 |9   |  14.764262| 117.447471|  14.703558|  14.84658|
|ENSMUSG00000054455 |9   |  27.198455|   7.336424|  20.908490| 110.55064|
|ENSMUSG00000097554 |9   |  41.186050|  28.140130|  40.780764|  41.20384|
|ENSMUSG00000039542 |9   |  49.615275|  75.098477|  49.494008|  50.10869|
|ENSMUSG00000038738 |9   |  50.108691|   6.657000|  49.051211|  78.22393|
|ENSMUSG00000038086 |9   |  51.898259|   9.842327|  51.472947|  52.50449|
|ENSMUSG00000032305 |9   |  57.362395|  14.166389|  57.359763|  57.96257|
|ENSMUSG00000039542 |9   |  58.357965|  24.965759|  58.264792|  58.51857|
|ENSMUSG00000026637 |9   |  66.202095|   6.128876|  61.127900| 118.85930|
|ENSMUSG00000039542 |9   | 107.647626|   6.709880| 103.046146| 108.28455|
|ENSMUSG00000032305 |9   | 117.664006|   7.429687| 116.937976| 118.80569|
|ENSMUSG00000044795 |10  |  22.909158|   7.195228|  21.125750|  23.14018|
|ENSMUSG00000019945 |10  |  68.436555|  36.340070|  68.296752|  68.55129|
|ENSMUSG00000019945 |10  |  68.966184|  32.570583|  68.966184|  69.22692|
|ENSMUSG00000071519 |10  |  75.574081|   6.462753|  73.705382|  83.04843|
|ENSMUSG00000029321 |10  | 125.591103|   7.538025| 125.539735| 126.86601|
|ENSMUSG00000080968 |11  |  16.903986|   6.044069|   9.065479|  41.96490|
|ENSMUSG00000069910 |11  |  34.822220|  26.757661|  34.701511|  34.82222|
|ENSMUSG00000033389 |11  |  65.063246|   9.777574|  64.648612|  68.88967|
|ENSMUSG00000041189 |11  |  67.078257|   6.065475|  64.368932|  70.61624|
|ENSMUSG00000044795 |11  |  68.892129|  10.613848|  65.096598|  69.41636|
|ENSMUSG00000027722 |11  |  73.111862|   6.349757|  73.005652|  77.67011|
|ENSMUSG00000039542 |11  |  78.151256|   6.402459|   5.561303| 105.51410|
|ENSMUSG00000027718 |11  |  79.152432|   6.886785|  70.782062|  82.07630|
|ENSMUSG00000051452 |11  |  84.107102|  10.077904|  83.647144|  84.68618|
|ENSMUSG00000025557 |11  |  98.275118|   6.140239|  90.223909|  99.34041|
|ENSMUSG00000080383 |11  | 109.423954|  11.546496| 109.406840| 111.19709|
|ENSMUSG00000027487 |12  |   6.347016|   6.708630|   5.846368|  11.34456|
|ENSMUSG00000033373 |12  |  76.788953|  32.214804|  76.548097|  76.88841|
|ENSMUSG00000021127 |12  |  83.238468|   6.982385|  82.129546|  86.61000|
|ENSMUSG00000058420 |12  |  84.148297|   6.072393|  80.160035| 117.40074|
|ENSMUSG00000021271 |12  |  87.166517|   6.136045|  56.572766|  88.46608|
|ENSMUSG00000049307 |12  | 107.833410|   6.639038| 106.974113| 109.52447|
|ENSMUSG00000021271 |12  | 110.197950|  33.653196| 110.180714| 111.45223|
|ENSMUSG00000099095 |13  |  15.611917|   8.154411|  13.283182|  18.32901|
|ENSMUSG00000015112 |13  |  35.457506|   6.362942|  34.416849| 105.15802|
|ENSMUSG00000027487 |13  |  64.344624|   9.185658|  63.548321|  67.69218|
|ENSMUSG00000097283 |13  |  65.206454|  10.783729|  64.841763|  66.93469|
|ENSMUSG00000021676 |13  |  95.733655| 100.243082|  95.733655|  95.85903|
|ENSMUSG00000069910 |13  | 101.786348|   6.927969| 101.137104| 102.83022|
|ENSMUSG00000050854 |14  |  31.435803|   6.178650|  28.932629|  94.70510|
|ENSMUSG00000086341 |14  |  48.126291|   7.138599|  46.678372|  51.99293|
|ENSMUSG00000025557 |14  |  99.566266|   6.659106|  13.014489| 101.18782|
|ENSMUSG00000025557 |14  | 121.571691|  35.458616| 121.480418| 121.67373|
|ENSMUSG00000022292 |15  |  38.227976|   7.784003|  36.256559|  41.45437|
|ENSMUSG00000015112 |15  |  58.990004|   7.527785|  58.658042|  60.07672|
|ENSMUSG00000060487 |15  |  60.990009|   6.909902|  58.857152|  67.06317|
|ENSMUSG00000022292 |15  |  64.643518|   6.404510|  62.181647|  72.90648|
|ENSMUSG00000024085 |15  |  66.679768|   7.227881|  63.659243|  68.92043|
|ENSMUSG00000080968 |15  | 100.424363|  59.731554|  99.800294| 100.43483|
|ENSMUSG00000026637 |16  |  94.789233|   6.874482|   3.000000|  94.97204|
|ENSMUSG00000015112 |17  |  43.242630|   7.536525|  34.179599|  43.56276|
|ENSMUSG00000024085 |17  |  64.655048|  55.404509|  64.600441|  64.72895|
|ENSMUSG00000069910 |17  |  69.244748|   6.998414|  68.506902|  70.46603|
|ENSMUSG00000086219 |17  |  78.830855|   7.823507|  77.910515|  80.14648|
|ENSMUSG00000070980 |17  |  89.875434|   6.190791|  78.362313|  94.98443|
|ENSMUSG00000032305 |18  |  68.177134|   6.075548|  65.041330|  68.20743|
|ENSMUSG00000002797 |19  |  38.421678|   6.062762|  36.662639|  40.14056|
|ENSMUSG00000044795 |19  |  40.204814|   6.610261|  38.774165|  40.22262|
|ENSMUSG00000021676 |X   |  12.322085|   7.621003|  11.895188|  12.99006|
|ENSMUSG00000086219 |X   |  49.017962|   7.209342|  48.174569|  50.20412|
|ENSMUSG00000009079 |X   |  53.140129|   7.474459|  51.892232|  59.08406|
|ENSMUSG00000067878 |X   |  56.774258|  16.144863|  53.969865|  56.80714|
|ENSMUSG00000015112 |X   | 150.513256|  10.317359| 145.335742| 157.49709|
|ENSMUSG00000058420 |X   | 153.682008|   7.479411| 146.728544| 163.71076|
|ENSMUSG00000022292 |X   | 158.566592|   6.854800| 155.383862| 160.34933|
|ENSMUSG00000071519 |X   | 166.335626|   8.361242| 164.415283| 167.22935|

### QTL Peaks Figure


~~~
qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
~~~
{: .language-r}

<img src="../fig/rmd-06-qtl_heatmap-1.png" alt="plot of chunk qtl_heatmap" width="612" style="display: block; margin: auto;" />

> ## Challenge
> What do the qtl scans for all gene exression traits look like? *Note:* Don't worry, we've done the qtl scans for you!!!
> You can read in this file, `../data/gene.norm_qtl_all.genes.Rdata`, which are the `scan1` results for all gene expression traits. 
>
> > ## Solution
> > 
> > 
> > ~~~
> > load("../data/gene.norm_qtl_all.genes.Rdata")
> > 
> > lod_threshold = 6
> > peaks = find_peaks(scan1_output = qtl, 
> >                map = map, 
> >                threshold = lod_threshold, 
> >                peakdrop = 4, 
> >                prob = 0.95)
> > write_csv(peaks, "../results/gene.norm_qtl_all.genes_peaks.csv")
> > 
> > ## Heat Map
> > qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
> > ~~~
> > {: .language-r}
> {: .solution}
{: .challenge}
