---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 07-map-many-eqtls.md in _episodes_rmd/
title: "Mapping Many Gene Expression Traits"
teaching: 30
exercises: 30
questions:
- "How do I map many genes?"
objectives:
- "????"
keypoints:
- "????"
source: Rmd
---



### Load Libraries  


~~~
library(tidyverse)
library(qtl2)
library(qtl2convert)
library(GGally)
library(broom)
library(knitr)
library(corrplot)
library(RColorBrewer)
library(qtl2ggplot)

source("../code/gg_transcriptome_map.R")
source("../code/qtl_heatmap.R")
~~~
{: .language-r}

## Load Data


~~~
# expression data
load("../data/attie_DO500_expr.datasets.RData")

# mapping data
load("../data/attie_DO500_mapping.data.RData")

probs <- readRDS("../data/attie_DO500_genoprobs_qtlviewer_8state_69k.rds")

# phenotypes
load("../data/attie_DO500_clinical.phenotypes.RData")
~~~
{: .language-r}


### Expression Data


Lets check the distribution



~~~
hist(counts$ENSMUSG00000020679, main = "Hnf1b (counts)")
~~~
{: .language-r}

<img src="../fig/rmd-07-hist_untransformed-1.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" />

~~~
hist(norm$ENSMUSG00000020679, main = "Hnf1b (norm)")
~~~
{: .language-r}

<img src="../fig/rmd-07-hist_untransformed-2.png" alt="plot of chunk hist_untransformed" width="612" style="display: block; margin: auto;" />

These counts are normalised

### The Marker Map  

The marker map for each chromosome is stored in the `map` object. This is used to plot the LOD scores calculated at each marker during QTL mapping.  Here we are using the 69K grid marker file



### Genotype probabilities  

We have already calculated genotype probabilities which we load above



### Covariates    

Now lets add the necessary covariates. For these analysis, lets see which covariates are significant (???)


~~~
# convert sex and DO wave (batch) to factors
pheno_clin$sex = factor(pheno_clin$sex)
pheno_clin$DOwave = factor(pheno_clin$DOwave)

covar = model.matrix(~sex + DOwave, data = pheno_clin)
~~~
{: .language-r}
### [Performing a genome scan](https://smcclatchy.github.io/mapping/06-perform-genome-scan/) 

Now lets perform the genome scan!


Pick 50 random genes (cis and trans genes)


### QTL Scans


~~~
genes = colnames(counts)
chr11 = which(genes=="ENSMUSG00000020679")
genes = genes[-chr11]

sams <- sample(length(genes), 50, replace = FALSE, prob = NULL)
genes <- genes[c(chr11, sams)]

qtl.file = "../data/gene.counts_qtl_cis.trans_random.rds"

qtl = NULL

if(file.exists(qtl.file)) {
  qtl = readRDS(qtl.file)
  } else {
    qtl = scan1(genoprobs = probs, 
                pheno = counts[,genes, drop = FALSE],
                kinship = K, 
                addcovar = covar, 
                cores = 2)
    saveRDS(qtl, file = qtl.file)
    }
~~~
{: .language-r}

### QTL plots


~~~
for(i in 1:ncol(qtl)) {
  plot_scan1(x = qtl, 
             map = map, 
             lodcolumn = i, 
             main = colnames(qtl)[i])
  abline(h = 6, col = 2, lwd = 2)
  }
~~~
{: .language-r}

<img src="../fig/rmd-07-qtl_plots-1.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-2.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-3.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-4.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-5.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-6.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-7.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-8.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-9.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-10.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-11.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-12.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-13.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-14.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-15.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-16.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-17.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-18.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-19.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-20.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-21.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-22.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-23.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-24.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-25.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-26.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-27.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-28.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-29.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-30.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-31.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-32.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-33.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-34.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-35.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-36.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-37.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-38.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-39.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-40.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-41.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-42.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-43.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-44.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-45.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-46.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-47.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-48.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-49.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-50.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" /><img src="../fig/rmd-07-qtl_plots-51.png" alt="plot of chunk qtl_plots" width="612" style="display: block; margin: auto;" />

### QTL Peaks


~~~
lod_threshold = 6
peaks = find_peaks(scan1_output = qtl, 
                   map = map, 
                   threshold = lod_threshold, 
                   peakdrop = 4, 
                   prob = 0.95)
kable(peaks %>% 
        dplyr::select(-lodindex) %>% 
        arrange(chr, pos), caption = "Expression QTL (eQTL) Peaks with LOD >= 6")
~~~
{: .language-r}



Table: Expression QTL (eQTL) Peaks with LOD >= 6

|lodcolumn          |chr |        pos|       lod|      ci_lo|      ci_hi|
|:------------------|:---|----------:|---------:|----------:|----------:|
|ENSMUSG00000089706 |1   |  99.660774| 13.502006|  97.226424| 110.897580|
|ENSMUSG00000022809 |1   | 156.039141|  7.356790| 155.767543| 161.069315|
|ENSMUSG00000040818 |1   | 158.378196|  6.025379| 154.582348| 192.703960|
|ENSMUSG00000027843 |1   | 190.869186|  7.670515| 154.502246| 190.893897|
|ENSMUSG00000026817 |2   |  32.649866| 14.659808|  32.582163|  32.877720|
|ENSMUSG00000035540 |2   |  71.207823| 10.033787|  69.444054|  72.066637|
|ENSMUSG00000027642 |2   |  71.876558|  6.566259|  69.446985|  72.228259|
|ENSMUSG00000024131 |2   |  72.186134|  6.653517|  70.133450|  74.887168|
|ENSMUSG00000075273 |2   |  75.982341| 39.697645|  75.783456|  76.020383|
|ENSMUSG00000031997 |2   |  77.126116|  6.609217|  32.582163| 172.256195|
|ENSMUSG00000075273 |2   |  80.269164| 16.405749|  79.904939|  80.276354|
|ENSMUSG00000074359 |2   | 123.785838| 19.075973| 121.238791| 124.109228|
|ENSMUSG00000027642 |2   | 157.011223| 15.309787| 156.938788| 157.983121|
|ENSMUSG00000032850 |2   | 165.283031|  6.794505| 156.592548| 167.545144|
|ENSMUSG00000026817 |2   | 165.711827|  6.804045| 162.899169| 167.033011|
|ENSMUSG00000022450 |2   | 166.613233|  6.086104|  23.160434| 172.756074|
|ENSMUSG00000081574 |2   | 176.031859| 12.283988| 174.924525| 178.092570|
|ENSMUSG00000024131 |2   | 180.569517|  7.133514| 180.122688| 182.107670|
|ENSMUSG00000039193 |2   | 181.340625|  6.549652| 180.191579| 182.107670|
|ENSMUSG00000074359 |3   |  30.314693|  6.238712|  19.044683|  88.287911|
|ENSMUSG00000030062 |3   |  56.483183|  7.490294|  55.225874|  58.529449|
|ENSMUSG00000027954 |3   |  89.539720| 12.406852|  87.759041|  90.434347|
|ENSMUSG00000032850 |3   |  90.438761|  6.241759|  87.757132|  99.567651|
|ENSMUSG00000070392 |3   |  95.852725|  7.499455|  93.224823| 100.725888|
|ENSMUSG00000027849 |3   | 103.057311| 16.830426| 102.096039| 103.531174|
|ENSMUSG00000073888 |4   |  42.112804| 68.597486|  41.822550|  42.403057|
|ENSMUSG00000086529 |4   |  95.746864|  6.917489|  94.573139|  97.521712|
|ENSMUSG00000070392 |4   | 118.988063|  6.756324| 117.435345| 123.347784|
|ENSMUSG00000024097 |4   | 119.286985|  6.053898|  40.808451| 121.948437|
|ENSMUSG00000062421 |4   | 123.325375|  6.442427| 120.342832| 123.347784|
|ENSMUSG00000081574 |4   | 149.936361|  6.244010| 132.893819| 151.750844|
|ENSMUSG00000097450 |5   |  75.941633|  7.533710|  66.516951|  86.022750|
|ENSMUSG00000032850 |5   | 118.286435| 42.184682| 117.795029| 118.298823|
|ENSMUSG00000056394 |5   | 148.040918|  6.025464|   3.590813| 151.833620|
|ENSMUSG00000027849 |6   |   4.770086|  8.365673|   4.330478|  13.311834|
|ENSMUSG00000027954 |6   |  18.399017|  6.139693|  17.565400|  22.701382|
|ENSMUSG00000039347 |6   |  32.525679|  6.643189|  29.688267|  33.138612|
|ENSMUSG00000029924 |6   |  37.782019| 15.482227|  37.581007|  39.640042|
|ENSMUSG00000039347 |6   |  48.206108| 18.732237|  47.270697|  48.979756|
|ENSMUSG00000030062 |6   |  87.749152|  6.658281|  72.922178|  90.840253|
|ENSMUSG00000002835 |6   | 143.002009|  6.184663| 128.530829| 144.204865|
|ENSMUSG00000085077 |6   | 147.783686| 10.981141| 147.623635| 148.631282|
|ENSMUSG00000056394 |7   |  11.723611| 10.196269|   4.780503|  15.927476|
|ENSMUSG00000096934 |7   |  19.586003|  6.860385|  19.284040|  24.736693|
|ENSMUSG00000022450 |7   |  29.331034|  7.274535|  28.735967|  83.797382|
|ENSMUSG00000022623 |7   |  30.531596|  6.715818|  30.434681| 140.177559|
|ENSMUSG00000033355 |7   |  45.720922|  7.205756|  44.940291|  46.361790|
|ENSMUSG00000035211 |7   |  99.647975| 14.595128|  99.345653| 101.659107|
|ENSMUSG00000002835 |7   | 125.066011|  6.596054|  79.171609| 126.089386|
|ENSMUSG00000081565 |7   | 138.649019|  6.122070| 137.253347| 139.107556|
|ENSMUSG00000020901 |8   |  13.394105|  6.524420|  11.247343|  14.658503|
|ENSMUSG00000002835 |8   | 110.699524|  7.943581| 110.246907| 112.860380|
|ENSMUSG00000075265 |9   |   3.000000|  6.444281|   3.000000|   5.338642|
|ENSMUSG00000031997 |9   |   9.153435| 15.585575|   7.737030|   9.630165|
|ENSMUSG00000039193 |9   |  13.830376|  6.165670|   9.230131|  21.177604|
|ENSMUSG00000040354 |9   |  48.786307|  6.588162|  47.590914|  49.372741|
|ENSMUSG00000040818 |9   |  49.100406|  6.584253|  47.541530|  50.281877|
|ENSMUSG00000020089 |9   |  99.812810|  6.179892|   9.204566| 105.783741|
|ENSMUSG00000033355 |9   |  99.961738|  6.101867|  94.660917| 101.882805|
|ENSMUSG00000031197 |10  |  35.580317|  6.226002|  32.519048|  39.162102|
|ENSMUSG00000040354 |10  | 127.550475|  7.888362| 127.014380| 127.603538|
|ENSMUSG00000085077 |11  |  48.841752|  6.055133|  47.191976|  49.366364|
|ENSMUSG00000025370 |11  |  49.307545|  6.247971|  46.739105|  96.907543|
|ENSMUSG00000020089 |11  |  69.829176|  7.270764|  69.234230|  74.716248|
|ENSMUSG00000033355 |11  |  69.829176| 10.301051|  69.810287|  71.782692|
|ENSMUSG00000002835 |11  |  69.829176|  6.151513|  54.389937| 105.507343|
|ENSMUSG00000034664 |11  | 102.764944|  6.751768| 102.148307| 103.726010|
|ENSMUSG00000062421 |11  | 103.863242|  7.734885| 102.711345| 104.979219|
|ENSMUSG00000024131 |12  |   4.421926|  6.817112|   3.437198|   6.012917|
|ENSMUSG00000027843 |12  | 108.590642|  8.906822| 105.606210| 108.930724|
|ENSMUSG00000025370 |12  | 108.722686|  6.374677| 107.835528| 111.453851|
|ENSMUSG00000031273 |12  | 113.140500|  9.030873| 109.597449| 113.596077|
|ENSMUSG00000078838 |13  |  37.811408|  6.404108|  37.640977|  42.288485|
|ENSMUSG00000021133 |13  | 112.665334|  6.292430|   3.000000| 113.812449|
|ENSMUSG00000022809 |13  | 112.978421|  6.151970| 110.627971| 115.753929|
|ENSMUSG00000097450 |14  |   4.551935| 44.794018|   3.000000|  12.527336|
|ENSMUSG00000097450 |14  |  19.690586| 43.179587|  19.646447|  19.702928|
|ENSMUSG00000097450 |14  |  19.783656| 41.918208|  19.772677|  20.178707|
|ENSMUSG00000097450 |14  |  20.957514| 34.880591|  20.940587|  21.161151|
|ENSMUSG00000040818 |14  |  22.328282|  8.499089|  21.072395|  27.181077|
|ENSMUSG00000022450 |14  |  25.735587|  6.219718|  23.768021|  82.208485|
|ENSMUSG00000031273 |14  |  58.622127|  6.317442|  47.047943|  60.660896|
|ENSMUSG00000040354 |14  |  91.235840|  7.559788|  83.327726|  93.259019|
|ENSMUSG00000035211 |14  |  94.472669|  6.386994|  87.138027|  99.705803|
|ENSMUSG00000025370 |15  |  27.402870|  6.188950|  13.897119|  60.990009|
|ENSMUSG00000078838 |15  |  58.506357|  8.348416|  53.666480|  58.980947|
|ENSMUSG00000022450 |15  |  68.313152|  9.957515|  68.064856|  71.555326|
|ENSMUSG00000022450 |15  |  82.378316| 15.532398|  81.625674|  82.387529|
|ENSMUSG00000097450 |15  |  93.641954|  6.772095|  77.426124|  94.143690|
|ENSMUSG00000097450 |16  |  10.457509|  7.002931|   9.864659|  10.764934|
|ENSMUSG00000038127 |16  |  26.926419| 19.126121|  26.087386|  27.189865|
|ENSMUSG00000022809 |16  |  27.544894|  6.689849|   4.509112|  28.539416|
|ENSMUSG00000075265 |16  |  30.485024| 11.304158|  30.284734|  33.180182|
|ENSMUSG00000022809 |16  |  38.391015| 71.452603|  38.137342|  38.398038|
|ENSMUSG00000034664 |16  |  74.569217|  6.001047|  70.573223|  76.101224|
|ENSMUSG00000089706 |16  |  92.782294|  8.061908|  92.274737|  93.569210|
|ENSMUSG00000078838 |17  |  26.806785| 17.985816|  26.579464|  27.452967|
|ENSMUSG00000074359 |17  |  33.542215|  9.828626|  32.772368|  34.982579|
|ENSMUSG00000057835 |17  |  55.910953| 14.549879|  55.628578|  56.231076|
|ENSMUSG00000002835 |17  |  56.830471| 14.700693|  55.530144|  57.071792|
|ENSMUSG00000039193 |17  |  75.380504| 19.345025|  74.966239|  75.607703|
|ENSMUSG00000039193 |17  |  79.552453|  9.416776|  79.487020|  79.958169|
|ENSMUSG00000096934 |18  |  32.496893|  7.085659|  28.934172|  34.041624|
|ENSMUSG00000044024 |18  |  38.025954| 21.012024|  37.796418|  38.136285|
|ENSMUSG00000034664 |19  |  36.366615|  6.558601|  36.083410|  37.221005|
|ENSMUSG00000022809 |X   |  13.518661|  8.634293|  13.517903|  16.380998|
|ENSMUSG00000035540 |X   |  53.220693|  8.200715|  41.456110|  61.164240|
|ENSMUSG00000086529 |X   |  74.373593|  6.617637|  73.635859|  75.096225|
|ENSMUSG00000070392 |X   | 139.419249|  7.533897| 136.453157| 142.824480|



~~~
write_csv(peaks, "../data/gene.counts_qtl_peaks_random.csv")
~~~
{: .language-r}

### QTL Peaks Figure


~~~
peaks = peaks %>%
  arrange(lodcolumn)
plot_peaks(peaks, 
           map, 
           col = c("blue","red"), 
           lwd = 3, 
           tick_height = 0.8, 
           gap = 0, 
           main = "LOD > 6")
box()
~~~
{: .language-r}

<img src="../fig/rmd-07-qtl_peaks_figure-1.png" alt="plot of chunk qtl_peaks_figure" width="612" style="display: block; margin: auto;" />

~~~
ggplot_peaks(peaks, 
             map, 
             col = c("blue","red"), 
             legend.title = "LOD > 6") 
~~~
{: .language-r}

<img src="../fig/rmd-07-qtl_peaks_figure-2.png" alt="plot of chunk qtl_peaks_figure" width="612" style="display: block; margin: auto;" />


~~~
qtl_heatmap(qtl = qtl, map = map, low.thr = 3.5)
~~~
{: .language-r}

<img src="../fig/rmd-07-qtl_heatmap-1.png" alt="plot of chunk qtl_heatmap" width="612" style="display: block; margin: auto;" />

